File: kb_modules/kb_typescript/postgres/data_structures/src/KBDataStructures.ts
import { KBSearch } from './KBSearch';
import { KBStatusData } from './KBStatusData';
import { KBJobQueue } from './KBJobQueue';
import { KBStream } from './KBStream';
import { KBRpcClient } from './KBRpcClient';
import { KBRpcServer } from './KBRpcServer';
import { KBLinkTable } from './KBLinkTable';
import { KBLinkMountTable } from './KBLinkMountTable';

export class KBDataStructures {
  public querySupport: KBSearch;

  // Delegated KB_Search methods
  public clearFilters: () => void;
  public searchLabel: (label: string) => void;
  public searchName: (name: string) => void;
  public searchPropertyKey: (key: string) => void;
  public searchPropertyValue: (key: string, value: any) => void;
  public searchHasLink: () => void;
  public searchHasLinkMount: () => void;
  public searchPath: (expr: string) => void;
  public searchStartingPath: (path: string) => void;
  public executeKbSearch: () => Promise<any[]>;
  public findDescription: (data: any) => any[];
  public findDescriptionPaths: (paths: string | string[]) => Promise<Record<string, any>>;
  public findPathValues: (rows: any[] | any) => string[];
  public decodeLinkNodes: (path: string) => [string, [string, string][]];

  // Status data
  public statusData: KBStatusData;
  public findStatusNodeIds: (...args: any[]) => Promise<any[]>;
  public findStatusNodeId: (...args: any[]) => Promise<any>;
  public getStatusData: (path: string) => Promise<[any, string]>;
  public setStatusData: (path: string, data: Record<string, any>) => Promise<[boolean, string]>;

  // Job queue
  public jobQueue: KBJobQueue;
  public findJobIds: (...args: any[]) => Promise<any[]>;
  public findJobId: (...args: any[]) => Promise<any>;
  public getQueuedNumber: (path: string) => Promise<number>;
  public getFreeNumber: (path: string) => Promise<number>;
  public peakJobData: (path: string, maxRetries?: number, retryDelay?: number) => Promise<any>;
  public markJobCompleted: (jobId: number, maxRetries?: number, retryDelay?: number) => Promise<any>;
  public pushJobData: (path: string, data: Record<string, any>, maxRetries?: number, retryDelay?: number) => Promise<any>;
  public listPendingJobs: (path: string, limit?: number, offset?: number) => Promise<any[]>;
  public listActiveJobs: (path: string, limit?: number, offset?: number) => Promise<any[]>;
  public clearJobQueue: (path: string) => Promise<any>;

  // Stream
  public stream: KBStream;
  public findStreamIds: (...args: any[]) => Promise<any[]>;
  public findStreamId: (...args: any[]) => Promise<any>;
  public findStreamTableKeys: (rows: any[]) => string[];
  public pushStreamData: (path: string, data: Record<string, any>, maxRetries?: number, retryDelay?: number) => Promise<any>;
  public listStreamData: (...args: any[]) => Promise<any[]>;
  public clearStreamData: (path: string, olderThan?: Date) => Promise<any>;
  public getStreamDataCount: (path: string, includeInvalid?: boolean) => Promise<number>;
  public getStreamDataRange: (path: string, start: Date, end: Date) => Promise<any[]>;
  public getStreamStatistics: (path: string, includeInvalid?: boolean) => Promise<Record<string, any>>;
  public getStreamDataById: (id: number) => Promise<any>;

  // RPC client
  public rpcClient: KBRpcClient;
  public rpcClientFindRpcClientId: (...args: any[]) => Promise<any>;
  public rpcClientFindRpcClientIds: (...args: any[]) => Promise<any[]>;
  public rpcClientFindRpcClientKeys: (data: any[]) => string[];
  public rpcClientFindFreeSlots: (clientPath: string) => Promise<number>;
  public rpcClientFindQueuedSlots: (clientPath: string) => Promise<number>;
  public rpcClientPeakAndClaimReplyData: (clientPath: string, maxRetries?: number, retryDelay?: number) => Promise<any>;
  public rpcClientClearReplyQueue: (clientPath: string, maxRetries?: number, retryDelay?: number) => Promise<number>;
  public rpcClientPushAndClaimReplyData: (...args: any[]) => Promise<void>;
  public rpcClientListWaitingJobs: (clientPath?: string) => Promise<any[]>;

  // RPC server
  public rpcServer: KBRpcServer;
  public rpcServerFindId: (...args: any[]) => Promise<any>;
  public rpcServerFindIds: (...args: any[]) => Promise<any[]>;
  public rpcServerFindTableKeys: (data: any[]) => string[];
  public rpcServerListJobsJobTypes: (serverPath: string, state: string) => Promise<any[]>;
  public rpcServerCountAllJobs: (serverPath: string) => Promise<any>;
  public rpcServerCountEmptyJobs: (serverPath: string) => Promise<number>;
  public rpcServerCountNewJobs: (serverPath: string) => Promise<number>;
  public rpcServerCountProcessingJobs: (serverPath: string) => Promise<number>;
  public rpcServerCountJobsJobTypes: (serverPath: string, state: string) => Promise<number>;
  public rpcServerPushRpcQueue: (...args: any[]) => Promise<any>;
  public rpcServerPeakServerQueue: (serverPath: string, retries?: number, waitTime?: number) => Promise<any>;
  public rpcServerMarkJobCompletion: (...args: any[]) => Promise<any>;
  public rpcServerClearServerQueue: (serverPath: string, maxRetries?: number, retryDelay?: number) => Promise<number>;

  // Link tables
  public linkTable: KBLinkTable;
  public linkTableFindRecordsByLinkName: (linkName: string, kb?: string) => Promise<any[]>;
  public linkTableFindRecordsByNodePath: (nodePath: string, kb?: string) => Promise<any[]>;
  public linkTableFindAllLinkNames: () => Promise<string[]>;
  public linkTableFindAllNodeNames: () => Promise<string[]>;

  public linkMountTable: KBLinkMountTable;
  public linkMountTableFindRecordsByLinkName: (linkName: string, kb?: string) => Promise<any[]>;
  public linkMountTableFindRecordsByMountPath: (mountPath: string, kb?: string) => Promise<any[]>;
  public linkMountTableFindAllLinkNames: () => Promise<string[]>;
  public linkMountTableFindAllMountPaths: () => Promise<string[]>;

  constructor(
    host: string,
    port: number,
    dbName: string,
    user: string,
    password: string,
    database: string
  ) {
    // Core search
    this.querySupport = new KBSearch(host, port, dbName, user, password, database);
    this.clearFilters = this.querySupport.clearFilters.bind(this.querySupport);
    this.searchLabel = this.querySupport.searchLabel.bind(this.querySupport);
    this.searchName = this.querySupport.searchName.bind(this.querySupport);
    this.searchPropertyKey = this.querySupport.searchPropertyKey.bind(this.querySupport);
    this.searchPropertyValue = this.querySupport.searchPropertyValue.bind(this.querySupport);
    this.searchHasLink = this.querySupport.searchHasLink.bind(this.querySupport);
    this.searchHasLinkMount = this.querySupport.searchHasLinkMount.bind(this.querySupport);
    this.searchPath = this.querySupport.searchPath.bind(this.querySupport);
    this.searchStartingPath = this.querySupport.searchStartingPath.bind(this.querySupport);
    this.executeKbSearch = this.querySupport.executeQuery.bind(this.querySupport);
    this.findDescription = this.querySupport.findDescription.bind(this.querySupport);
    this.findDescriptionPaths = this.querySupport.findDescriptionPaths.bind(this.querySupport);
    this.findPathValues = this.querySupport.findPathValues.bind(this.querySupport);
    this.decodeLinkNodes = this.querySupport.decodeLinkNodes.bind(this.querySupport);

    // Status Data
    this.statusData = new KBStatusData(this.querySupport, database);
    this.findStatusNodeIds = this.statusData.findNodeIds.bind(this.statusData);
    this.findStatusNodeId = this.statusData.findNodeId.bind(this.statusData);
    this.getStatusData = this.statusData.getStatusData.bind(this.statusData);
    this.setStatusData = this.statusData.setStatusData.bind(this.statusData);

    // Job Queue
    this.jobQueue = new KBJobQueue(this.querySupport, database);
    this.findJobIds = this.jobQueue.findJobIds.bind(this.jobQueue);
    this.findJobId = this.jobQueue.findJobId.bind(this.jobQueue);
    this.getQueuedNumber = this.jobQueue.getQueuedNumber.bind(this.jobQueue);
    this.getFreeNumber = this.jobQueue.getFreeNumber.bind(this.jobQueue);
    this.peakJobData = this.jobQueue.peakJobData.bind(this.jobQueue);
    this.markJobCompleted = this.jobQueue.markJobCompleted.bind(this.jobQueue);
    this.pushJobData = this.jobQueue.pushJobData.bind(this.jobQueue);
    this.listPendingJobs = this.jobQueue.listPendingJobs.bind(this.jobQueue);
    this.listActiveJobs = this.jobQueue.listActiveJobs.bind(this.jobQueue);
    this.clearJobQueue = this.jobQueue.clearJobQueue.bind(this.jobQueue);

    // Stream
    this.stream = new KBStream(this.querySupport, database);
    this.findStreamIds = this.stream.findStreamIds.bind(this.stream);
    this.findStreamId = this.stream.findStreamId.bind(this.stream);
    this.findStreamTableKeys = this.stream.findStreamTableKeys.bind(this.stream);
    this.pushStreamData = this.stream.pushStreamData.bind(this.stream);
    this.listStreamData = this.stream.listStreamData.bind(this.stream);
    this.clearStreamData = this.stream.clearStreamData.bind(this.stream);
    this.getStreamDataCount = this.stream.getStreamDataCount.bind(this.stream);
    this.getStreamDataRange = this.stream.getStreamDataRange.bind(this.stream);
    this.getStreamStatistics = this.stream.getStreamStatistics.bind(this.stream);
    this.getStreamDataById = this.stream.getStreamDataById.bind(this.stream);

    // RPC Client
    this.rpcClient = new KBRpcClient(this.querySupport, database);
    this.rpcClientFindRpcClientId = this.rpcClient.findRpcClientId.bind(this.rpcClient);
    this.rpcClientFindRpcClientIds = this.rpcClient.findRpcClientIds.bind(this.rpcClient);
    this.rpcClientFindRpcClientKeys = this.rpcClient.findRpcClientKeys.bind(this.rpcClient);
    this.rpcClientFindFreeSlots = this.rpcClient.findFreeSlots.bind(this.rpcClient);
    this.rpcClientFindQueuedSlots = this.rpcClient.findQueuedSlots.bind(this.rpcClient);
    this.rpcClientPeakAndClaimReplyData = this.rpcClient.peakAndClaimReplyData.bind(this.rpcClient);
    this.rpcClientClearReplyQueue = this.rpcClient.clearReplyQueue.bind(this.rpcClient);
    this.rpcClientPushAndClaimReplyData = this.rpcClient.pushAndClaimReplyData.bind(this.rpcClient);
    this.rpcClientListWaitingJobs = this.rpcClient.listWaitingJobs.bind(this.rpcClient);

    // RPC Server
    this.rpcServer = new KBRpcServer(this.querySupport, database);
    this.rpcServerFindId = this.rpcServer.findRpcServerId.bind(this.rpcServer);
    this.rpcServerFindIds = this.rpcServer.findRpcServerIds.bind(this.rpcServer);
    this.rpcServerFindTableKeys = this.rpcServer.findRpcServerTableKeys.bind(this.rpcServer);
    this.rpcServerListJobsJobTypes = this.rpcServer.listJobsJobTypes.bind(this.rpcServer);
    this.rpcServerCountAllJobs = this.rpcServer.countAllJobs.bind(this.rpcServer);
    this.rpcServerCountEmptyJobs = this.rpcServer.countEmptyJobs.bind(this.rpcServer);
    this.rpcServerCountNewJobs = this.rpcServer.countNewJobs.bind(this.rpcServer);
    this.rpcServerCountProcessingJobs = this.rpcServer.countProcessingJobs.bind(this.rpcServer);
    this.rpcServerCountJobsJobTypes = this.rpcServer.countJobsJobTypes.bind(this.rpcServer);
    this.rpcServerPushRpcQueue = this.rpcServer.pushRpcQueue.bind(this.rpcServer);
    this.rpcServerPeakServerQueue = this.rpcServer.peakServerQueue.bind(this.rpcServer);
    this.rpcServerMarkJobCompletion = this.rpcServer.markJobCompletion.bind(this.rpcServer);
    this.rpcServerClearServerQueue = this.rpcServer.clearServerQueue.bind(this.rpcServer);

    // Link Table
    this.linkTable = new KBLinkTable(this.querySupport.getConn(), database);
    this.linkTableFindRecordsByLinkName = this.linkTable.findRecordsByLinkName.bind(this.linkTable);
    this.linkTableFindRecordsByNodePath = this.linkTable.findRecordsByNodePath.bind(this.linkTable);
    this.linkTableFindAllLinkNames = this.linkTable.findAllLinkNames.bind(this.linkTable);
    this.linkTableFindAllNodeNames = this.linkTable.findAllNodeNames.bind(this.linkTable);

    // Link Mount Table
    this.linkMountTable = new KBLinkMountTable(this.querySupport.getConn(), database);
    this.linkMountTableFindRecordsByLinkName = this.linkMountTable.findRecordsByLinkName.bind(this.linkMountTable);
    this.linkMountTableFindRecordsByMountPath = this.linkMountTable.findRecordsByMountPath.bind(this.linkMountTable);
    this.linkMountTableFindAllLinkNames = this.linkMountTable.findAllLinkNames.bind(this.linkMountTable);
    this.linkMountTableFindAllMountPaths = this.linkMountTable.findAllMountPaths.bind(this.linkMountTable);
  }
}

File: kb_modules/kb_go/postgres/data_structures/data_structures_module/KBDataStructures.go
package data_structures_module

import (
	//database/sql"
	"fmt"
	//"log"
	//"time"

	//"github.com/google/uuid"
	_ "github.com/lib/pq" // PostgreSQL driver
)

// KBDataStructures handles the data structures for the knowledge base
type KBDataStructures struct {
	// Core components
	querySupport    *KBSearch
	statusData      *KBStatusData
	jobQueue        *KBJobQueue
	stream          *KBStream
	rpcClient       *KBRPCClient
	rpcServer       *KBRPCServer
	linkTable       *KBLinkTable
	linkMountTable  *KBLinkMountTable
}

// NewKBDataStructures creates a new instance of KBDataStructures
func NewKBDataStructures(host, port, dbname, user, password, database string) (*KBDataStructures, error) {
	// Initialize the query support (equivalent to KB_Search)
	querySupport, err := NewKBSearch(host, port, dbname, user, password, database)
	if err != nil {
		return nil, fmt.Errorf("failed to create KB_Search: %w", err)
	}

	// Initialize all components
	statusData := NewKBStatusData(querySupport, database)
	jobQueue := NewKBJobQueue(querySupport, database)
	stream := NewKBStream(querySupport, database)
	rpcClient := NewKBRPCClient(querySupport, database)
	rpcServer := NewKBRPCServer(querySupport, database)
	linkTable := NewKBLinkTable(querySupport.conn, database)
	linkMountTable := NewKBLinkMountTable(querySupport.conn, database)

	return &KBDataStructures{
		querySupport:   querySupport,
		statusData:     statusData,
		jobQueue:       jobQueue,
		stream:         stream,
		rpcClient:      rpcClient,
		rpcServer:      rpcServer,
		linkTable:      linkTable,
		linkMountTable: linkMountTable,
	}, nil
}
/*
// Query Support Methods (delegated to querySupport)
func (kds *KBDataStructures) ClearFilters() {
	kds.querySupport.ClearFilters()
}

func (kds *KBDataStructures) SearchLabel(label string) {
	kds.querySupport.SearchLabel(label)
}

func (kds *KBDataStructures) SearchName(name string) {
	kds.querySupport.SearchName(name)
}

func (kds *KBDataStructures) SearchPropertyKey(key string) {
	kds.querySupport.SearchPropertyKey(key)
}

func (kds *KBDataStructures) SearchPropertyValue(value string,property_value map[string]interface{}) {
	kds.querySupport.SearchPropertyValue(value,property_value)
}

func (kds *KBDataStructures) SearchHasLink() {
	kds.querySupport.SearchHasLink()
}

func (kds *KBDataStructures) SearchHasLinkMount() {
	kds.querySupport.SearchHasLinkMount()
}


func (kds *KBDataStructures) SearchPath(path string) {
	kds.querySupport.SearchPath(path)
}

func (kds *KBDataStructures) SearchStartingPath(path string) {
	kds.querySupport.SearchStartingPath(path)
}

func (kds *KBDataStructures) ExecuteKBSearch(property_value map[string]interface{}) ([]map[string]interface{}, error) {
	return kds.querySupport.ExecuteQuery()
}

func (kds *KBDataStructures) FindDescription(row map[string]interface{}) map[string]string {
	return kds.querySupport.FindDescription(row)
}
func (kds *KBDataStructures) FindDescriptions(row []map[string]interface{}) []map[string]string {
	return kds.querySupport.FindDescriptions(row)
}


func (kds *KBDataStructures) FindDescriptionPaths(paths []string) ([]map[string]interface{}, error) {
	return kds.querySupport.FindDescriptionPaths(paths)
}

func (kds *KBDataStructures) FindDescriptionPath(path string) (map[string]interface{}, error) {
	return kds.querySupport.FindDescriptionPath(path)
}

func (kds *KBDataStructures) FindPathValues(keyData []map[string]interface{}) ([]string) {
	return kds.querySupport.FindPathValues(keyData)
}

func (kds *KBDataStructures) DecodeLinkNodes(path string) (string, [][]string, error) {
	return kds.querySupport.DecodeLinkNodes(path)
}








// Status Data Methods (delegated to statusData)
func (kds *KBDataStructures) FindStatusNodeIDs(kb, nodeName *string, properties map[string]interface{}, nodePath *string) ([]map[string]interface{}, error) {
	return kds.statusData.FindNodeIDs(kb, nodeName, properties, nodePath)
}



func (kds *KBDataStructures) GetStatusData(path string) (map[string]interface{},string, error) {
	return kds.statusData.GetStatusData(path)
}

func (kds *KBDataStructures) SetStatusData(path string, data map[string]interface{},retryCount int, retryDelay time.Duration) (bool ,string,error){
	return kds.statusData.SetStatusData(path, data,retryCount, retryDelay)
}

// Job Queue Methods (delegated to jobQueue)
func (kds *KBDataStructures) FindJobID(kb *string, nodeName *string, properties map[string]interface{}, nodePath *string) (map[string]interface{}, error) {
	return kds.jobQueue.FindJobID(kb, nodeName, properties, nodePath)
}
func (kds *KBDataStructures) FindJobIDs(kb *string, nodeName *string, properties map[string]interface{}, nodePath *string) ([]map[string]interface{}, error) {
	return kds.jobQueue.FindJobIDs(kb, nodeName, properties, nodePath)
}


func (kds *KBDataStructures) GetQueuedNumber(jobPath string) (int, error) {
	return kds.jobQueue.GetQueuedNumber(jobPath)
}

func (kds *KBDataStructures) GetFreeNumber(jobPath string) (int, error) {
	return kds.jobQueue.GetFreeNumber(jobPath)
}

func (kds *KBDataStructures) PeakJobData(jobPath string, maxRetries int, retryDelay time.Duration) (*PeakJobResult,error) {
	return kds.jobQueue.PeakJobData(jobPath, maxRetries, retryDelay)
}

func (kds *KBDataStructures) MarkJobCompleted(jobID int, maxRetries int, retryDelay time.Duration) (*JobCompletionResult, error) {
	return kds.jobQueue.MarkJobCompleted(jobID, maxRetries, retryDelay)
}

func (kds *KBDataStructures) PushJobData(jobPath string, data map[string]interface{}, maxRetries int, retryDelay time.Duration) (*PushJobResult, error) {
	return kds.jobQueue.PushJobData(jobPath, data, maxRetries, retryDelay)
}

func (kds *KBDataStructures) ListPendingJobs(jobPath string, limit *int, offset int) ([]JobRecord, error) {
	return kds.jobQueue.ListPendingJobs(jobPath, limit, offset)
}

func (kds *KBDataStructures) ListActiveJobs(jobPath string, limit *int, offset int) ([]JobRecord, error) {
	return kds.jobQueue.ListActiveJobs(jobPath, limit, offset)
}

func (kds *KBDataStructures) ClearJobQueue(jobPath string) (*ClearQueueResult, error) {
	return kds.jobQueue.ClearJobQueue(jobPath)
}
func (kds *KBDataStructures) FindStreamIDs(kb *string, nodeName *string, properties map[string]interface{}, nodePath *string) ([]map[string]interface{}, error) {
	return kds.stream.FindStreamIDs(kb, nodeName, properties, nodePath)
}
// Stream Methods (delegated to stream)


func (kds *KBDataStructures) FindStreamID(kb *string, nodeName *string, properties map[string]interface{}, nodePath *string) (map[string]interface{}, error) {
	return kds.stream.FindStreamID(kb, nodeName, properties, nodePath)
}

func (kds *KBDataStructures) FindStreamTableKeys(nodeIDs []map[string]interface{}) ([]string) {
	return kds.stream.FindStreamTableKeys(nodeIDs)
}

func (kds *KBDataStructures) PushStreamData(streamKey string, data map[string]interface{}, maxRetries int, retryDelay time.Duration) (*StreamPushResult, error) {
	return kds.stream.PushStreamData(streamKey, data, maxRetries, retryDelay)
}

func (kds *KBDataStructures) ListStreamData(path string, limit *int, offset int, recordedAfter, recordedBefore *time.Time, order string) ([]StreamRecord, error) {
	return kds.stream.ListStreamData(path, limit, offset, recordedAfter, recordedBefore, order)
}

func (kds *KBDataStructures) ClearStreamData(streamKey string) error {
	return kds.stream.ClearStreamData(streamKey)
}

func (kds *KBDataStructures) GetStreamDataCount(streamKey string) (int, error) {
	return kds.stream.GetStreamDataCount(streamKey)
}

func (kds *KBDataStructures) GetStreamDataRange(streamKey string, start, end int) ([]map[string]interface{}, error) {
	return kds.stream.GetStreamDataRange(streamKey, start, end)
}

func (kds *KBDataStructures) GetStreamStatistics(streamKey string) (map[string]interface{}, error) {
	return kds.stream.GetStreamStatistics(streamKey)
}

func (kds *KBDataStructures) GetStreamDataByID(streamKey string, id int) (map[string]interface{}, error) {
	return kds.stream.GetStreamDataByID(streamKey, id)
}

// RPC Client Methods (delegated to rpcClient)
func (kds *KBDataStructures) RPCClientFindRPCClientID(nodeName *string, properties map[string]interface{}, nodePath *string) ([]string, error) {
	return kds.rpcClient.FindRPCClientID(nodeName, properties, nodePath)
}

func (kds *KBDataStructures) RPCClientFindRPCClientIDs(nodeName *string, properties map[string]interface{}, nodePath *string) ([]string, error) {
	return kds.rpcClient.FindRPCClientIDs(nodeName, properties, nodePath)
}

func (kds *KBDataStructures) RPCClientFindRPCClientKeys(nodeIDs []string) ([]string, error) {
	return kds.rpcClient.FindRPCClientKeys(nodeIDs)
}

func (kds *KBDataStructures) RPCClientFindFreeSlots(clientPath string) (int, error) {
	return kds.rpcClient.FindFreeSlots(clientPath)
}

func (kds *KBDataStructures) RPCClientFindQueuedSlots(clientPath string) (int, error) {
	return kds.rpcClient.FindQueuedSlots(clientPath)
}

func (kds *KBDataStructures) RPCClientPeakAndClaimReplyData(clientPath string) (map[string]interface{}, error) {
	return kds.rpcClient.PeakAndClaimReplyData(clientPath)
}

func (kds *KBDataStructures) RPCClientClearReplyQueue(clientPath string) error {
	return kds.rpcClient.ClearReplyQueue(clientPath)
}

func (kds *KBDataStructures) RPCClientPushAndClaimReplyData(clientPath string, requestID interface{}, serverPath, action, transactionTag string, replyPayload map[string]interface{}) error {
	return kds.rpcClient.PushAndClaimReplyData(clientPath, requestID, serverPath, action, transactionTag, replyPayload)
}

func (kds *KBDataStructures) RPCClientListWaitingJobs(clientPath string) ([]map[string]interface{}, error) {
	return kds.rpcClient.ListWaitingJobs(clientPath)
}

// RPC Server Methods (delegated to rpcServer)
func (kds *KBDataStructures) RPCServerIDFind(nodeName *string, properties map[string]interface{}, nodePath *string) ([]string, error) {
	return kds.rpcServer.FindRPCServerID(nodeName, properties, nodePath)
}

func (kds *KBDataStructures) RPCServerIDsFind(nodeName *string, properties map[string]interface{}, nodePath *string) ([]string, error) {
	return kds.rpcServer.FindRPCServerIDs(nodeName, properties, nodePath)
}

func (kds *KBDataStructures) RPCServerTableKeysFind(nodeIDs []string) ([]string, error) {
	return kds.rpcServer.FindRPCServerTableKeys(nodeIDs)
}

func (kds *KBDataStructures) RPCServerListJobsJobTypes(serverPath, jobType string) ([]map[string]interface{}, error) {
	return kds.rpcServer.ListJobsJobTypes(serverPath, jobType)
}

func (kds *KBDataStructures) RPCServerCountAllJobs(serverPath string) (int, error) {
	return kds.rpcServer.CountAllJobs(serverPath)
}

func (kds *KBDataStructures) RPCServerCountEmptyJobs(serverPath string) (int, error) {
	return kds.rpcServer.CountEmptyJobs(serverPath)
}

func (kds *KBDataStructures) RPCServerCountNewJobs(serverPath string) (int, error) {
	return kds.rpcServer.CountNewJobs(serverPath)
}

func (kds *KBDataStructures) RPCServerCountProcessingJobs(serverPath string) (int, error) {
	return kds.rpcServer.CountProcessingJobs(serverPath)
}

func (kds *KBDataStructures) RPCServerCountJobsJobTypes(serverPath, jobType string) (int, error) {
	return kds.rpcServer.CountJobsJobTypes(serverPath, jobType)
}

func (kds *KBDataStructures) RPCServerPushRPCQueue(serverPath, requestID, rpcAction string, requestPayload map[string]interface{}, transactionTag string, priority int, rpcClientQueue string, maxRetries int, waitTime float64) error {
	return kds.rpcServer.PushRPCQueue(serverPath, requestID, rpcAction, requestPayload, transactionTag, priority, rpcClientQueue, maxRetries, waitTime)
}

func (kds *KBDataStructures) RPCServerPeakServerQueue(serverPath string) (map[string]interface{}, error) {
	return kds.rpcServer.PeakServerQueue(serverPath)
}

func (kds *KBDataStructures) RPCServerMarkJobCompletion(serverPath string, jobID interface{}) error {
	return kds.rpcServer.MarkJobCompletion(serverPath, jobID)
}

func (kds *KBDataStructures) RPCServerClearServerQueue(serverPath string) error {
	return kds.rpcServer.ClearServerQueue(serverPath)
}

// Link Table Methods (delegated to linkTable)
func (kds *KBDataStructures) LinkTableFindRecordsByLinkName(linkName string, kb *string) ([]map[string]interface{}, error) {
	return kds.linkTable.FindRecordsByLinkName(linkName, kb)
}

func (kds *KBDataStructures) LinkTableFindRecordsByNodePath(nodePath string, kb *string) ([]map[string]interface{}, error) {
	return kds.linkTable.FindRecordsByNodePath(nodePath, kb)
}

func (kds *KBDataStructures) LinkTableFindAllLinkNames() ([]string, error) {
	return kds.linkTable.FindAllLinkNames()
}

func (kds *KBDataStructures) LinkTableFindAllNodeNames() ([]string, error) {
	return kds.linkTable.FindAllNodeNames()
}

// Link Mount Table Methods (delegated to linkMountTable)
func (kds *KBDataStructures) LinkMountTableFindRecordsByLinkName(linkName string, kb *string) ([]map[string]interface{}, error) {
	return kds.linkMountTable.FindRecordsByLinkName(linkName, kb)
}

func (kds *KBDataStructures) LinkMountTableFindRecordsByMountPath(mountPath string, kb *string) ([]map[string]interface{}, error) {
	return kds.linkMountTable.FindRecordsByMountPath(mountPath, kb)
}

func (kds *KBDataStructures) LinkMountTableFindAllLinkNames() ([]string, error) {
	return kds.linkMountTable.FindAllLinkNames()
}

func (kds *KBDataStructures) LinkMountTableFindAllMountPaths() ([]string, error) {
	return kds.linkMountTable.FindAllMountPaths()
}

// Disconnect closes the database connection
func (kds *KBDataStructures) Disconnect() error {
	return kds.querySupport.Disconnect()
}
	*/
